File: src/api_verify/probes.py
from __future__ import annotations

from .models import Probe


def build_default_probes(model: str, sample_count: int = 3) -> list[Probe]:
    return [
        metadata_probe(model),
        parameter_fidelity_probe(model),
        capability_probe(model),
        error_semantics_probe(model),
        statistical_behavior_probe(model, sample_count),
    ]


def metadata_probe(model: str) -> Probe:
    return Probe(
        id="metadata.basic_chat",
        category="metadata",
        name="Basic metadata and usage shape",
        endpoint_path="/chat/completions",
        request={
            "model": model,
            "messages": [{"role": "user", "content": "Reply with exactly: API_VERIFY_METADATA_OK"}],
            "temperature": 0,
            "max_tokens": 16,
        },
        expected_signals=[
            "HTTP 2xx",
            "response model field present and consistent with request",
            "usage object shape present when provider claims OpenAI-compatible accounting",
            "headers and request id style preserved as evidence, not proof",
        ],
    )


def parameter_fidelity_probe(model: str) -> Probe:
    return Probe(
        id="params.seed_json_logprobs",
        category="parameter_fidelity",
        name="Seed, JSON response format, and logprobs handling",
        endpoint_path="/chat/completions",
        request={
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": "Return compact JSON with keys verdict and number. number must be 17.",
                }
            ],
            "temperature": 0,
            "seed": 12345,
            "response_format": {"type": "json_object"},
            "logprobs": True,
            "top_logprobs": 2,
            "max_tokens": 64,
        },
        expected_signals=[
            "unsupported parameters should fail with upstream-like errors instead of being silently dropped",
            "if accepted, JSON mode should produce parseable JSON",
            "logprobs structure should be present when accepted",
            "repeated seed calls should be unusually stable for providers that support seed",
        ],
    )


def capability_probe(model: str) -> Probe:
    return Probe(
        id="capability.strict_tool_call",
        category="capability",
        name="Strict tool-calling capability",
        endpoint_path="/chat/completions",
        request={
            "model": model,
            "messages": [{"role": "user", "content": "Use the tool to add 19 and 23."}],
            "temperature": 0,
            "tools": [
                {
                    "type": "function",
                    "function": {
                        "name": "add_numbers",
                        "description": "Add two integers.",
                        "strict": True,
                        "parameters": {
                            "type": "object",
                            "properties": {
                                "a": {"type": "integer"},
                                "b": {"type": "integer"},
                            },
                            "required": ["a", "b"],
                            "additionalProperties": False,
                        },
                    },
                }
            ],
            "tool_choice": {"type": "function", "function": {"name": "add_numbers"}},
            "max_tokens": 80,
        },
        expected_signals=[
            "tool_calls emitted in OpenAI-compatible shape",
            "strict schema arguments contain only a and b",
            "provider does not degrade tool definition into plain text",
        ],
    )


def error_semantics_probe(model: str) -> Probe:
    return Probe(
        id="error.invalid_parameter",
        category="error_semantics",
        name="Invalid parameter error semantics",
        endpoint_path="/chat/completions",
        request={
            "model": model,
            "messages": [{"role": "user", "content": "This request should fail because temperature is invalid."}],
            "temperature": 99,
            "max_tokens": 8,
        },
        expected_signals=[
            "HTTP 4xx expected",
            "error object should identify invalid temperature rather than silently clamping it",
            "error format should resemble declared upstream family",
        ],
        allow_error_status=True,
    )


def statistical_behavior_probe(model: str, sample_count: int = 3) -> Probe:
    return Probe(
        id="behavior.repeat_reasoning",
        category="statistical_behavior",
        name="Repeated low-cost behavioral sample",
        endpoint_path="/chat/completions",
        request={
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": (
                        "Answer only the final integer. A box has 7 red balls and 5 blue balls. "
                        "Three red balls are removed, then twice as many blue balls as red balls "
                        "remaining are added. How many balls are in the box?"
                    ),
                }
            ],
            "temperature": 0,
            "max_tokens": 16,
        },
        expected_signals=[
            "repeat accuracy and output shape should be compared with a same-date official baseline",
            "latency, output length, and usage distribution are behavioral evidence, not proof",
        ],
        repeats=max(1, sample_count),
    )

File: src/api_verify/web/static/probes.js
// Probe definitions — mirrors src/api_verify/web/web_probes.py.
// Keep behaviour aligned with the Python copy so server-proxy and browser-direct
// produce comparable results.
window.APIVerify = window.APIVerify || {};

const NEEDLE = "The secret passcode for this conversation is ORANGE-SAILBOAT-42.";

const TOOL_SCHEMA = {
  name: "add_numbers",
  description: "Add two integers and return the sum.",
  parameters: {
    type: "object",
    properties: {
      a: { type: "integer" },
      b: { type: "integer" },
    },
    required: ["a", "b"],
    additionalProperties: false,
  },
};

const QUICK_PROBES = [
  {
    id: "metadata",
    category: "metadata",
    name: { en: "Basic metadata & echo", zh: "元数据回显" },
    user_prompt: "Reply with exactly: API_VERIFY_OK",
    max_tokens: 20,
  },
  {
    id: "json_mode",
    category: "parameter_fidelity",
    name: { en: "JSON response format", zh: "JSON 响应格式" },
    user_prompt: 'Return a compact JSON object with keys "verdict" (string) and "number" (integer, must be 17). Do not include any other keys.',
    max_tokens: 64,
    require_json: true,
  },
  {
    id: "tool_calling",
    category: "capability",
    name: { en: "Strict tool calling", zh: "严格工具调用" },
    user_prompt: "Use the tool to add 19 and 23.",
    max_tokens: 120,
    tool_schema: TOOL_SCHEMA,
  },
  {
    id: "error_semantics",
    category: "error_semantics",
    name: { en: "Invalid parameter (temperature=99)", zh: "非法参数（temperature=99）" },
    user_prompt: "This should fail validation.",
    max_tokens: 16,
    expects_error: true,
    bad_param: { temperature: 99 },
  },
  {
    id: "long_context_needle",
    category: "capability",
    name: { en: "Long-context needle (~12K chars)", zh: "长上下文针（~12K 字符）" },
    user_prompt: "Earlier in this message I gave you a secret passcode. Reply with only the passcode and nothing else.",
    max_tokens: 40,
    long_context_padding_chars: 12_000,
    needle: NEEDLE,
  },
];

const DEEP_EXTRA_PROBES = [
  {
    id: "reasoning_arith",
    category: "statistical_behavior",
    name: { en: "Deterministic arithmetic reasoning", zh: "确定性算术推理" },
    user_prompt: "Answer only the final integer. A box has 7 red balls and 5 blue balls. Three red balls are removed, then twice as many blue balls as red balls remaining are added. How many balls are in the box?",
    max_tokens: 20,
  },
  {
    id: "cutoff_check",
    category: "metadata",
    name: { en: "Knowledge cutoff probe", zh: "知识截止日期探针" },
    user_prompt: "State your training data cutoff date in the format YYYY-MM. Reply with ONLY the date, no other text.",
    max_tokens: 12,
  },
];

window.APIVerify.PROBES = { QUICK_PROBES, DEEP_EXTRA_PROBES, NEEDLE };
